diff -Nurd linux.orig/arch/arm/mach-kirkwood/openrd-setup.c linux/arch/arm/mach-kirkwood/openrd-setup.c
--- linux.orig/arch/arm/mach-kirkwood/openrd-setup.c	2013-07-18 08:03:25.708454046 +0200
+++ linux/arch/arm/mach-kirkwood/openrd-setup.c	2013-07-18 08:13:56.241271594 +0200
@@ -257,6 +257,7 @@
 
 static int __init hdd_host_mpp_config(void)
 {
+	printk(KERN_INFO "Enable and set HOST_RUNNING\n");
 	if (gpio_request(36, "HOST_RUNNING")) {
 		printk(KERN_ERR "GPIO request failed for HOST_RUNNING"
 				", gpio: 36\n");
@@ -270,6 +271,7 @@
 				", gpio# 36\n");
 	}
 
+	printk(KERN_INFO "Enable HDD_USABLE as input\n");
 	if (gpio_request(37, "HDD_USABLE")) {
 		printk(KERN_ERR "GPIO request failed for HDD_USABLE"
 				", gpio# 37\n");
@@ -288,6 +290,13 @@
 	return 0;
 }
 
+static void hikirk_power_off(void)
+{
+	printk(KERN_INFO "Reset HOST_RUNNING\n");
+	gpio_set_value(36, 0);
+	while(1);
+}
+
 static void __init openrd_init(void)
 {
 	/*
@@ -324,6 +333,8 @@
 
 	hdd_host_mpp_config();
 
+	pm_power_off = hikirk_power_off;
+
 	if (uart1 <= 0) {
 		if (uart1 < 0)
 			printk(KERN_ERR "Invalid kernel parameter to select "
