diff --git a/drivers/mtd/spi/stmicro.c b/drivers/mtd/spi/stmicro.c
index a9b33cf..62b57d2 100644
--- a/drivers/mtd/spi/stmicro.c
+++ b/drivers/mtd/spi/stmicro.c
@@ -106,10 +106,10 @@ static const struct stmicro_spi_flash_params stmicro_spi_flash_table[] = {
 	},
 	{
 		.idcode1 = 0x18,
-		.page_size = 256,
-		.pages_per_sector = 1024,
-		.nr_sectors = 64,
-		.name = "M25P128",
+		.page_size = 256,
+		.pages_per_sector = 256,
+		.nr_sectors = 256,
+		.name = "N25Q128",
 	},
 };
 
